### PR TITLE
Add support for sideEffects glob(s)

### PR DESCRIPTION
**Pull request by reergymerej**
In addition to Booleans, the `sideEffects` value in a package can
identify child modules with side effects by a glob or Array of globs.

**What kind of change does this PR introduce?**

enhancement to `sideEffects` optimization

**Did you add tests for your changes?**

Yes, unit and integration.

**If relevant, link to documentation update:**

I have not created the PR for the docs yet, but I've got some guidance on where to do it.
https://github.com/webpack/webpack.js.org/issues/1706

**Summary**

This allows users more granularity when specifying if child modules introduce side effects.

This originated in #6074, but it _DOES NOT_  address the `"false"` option.  During a previous PR
https://github.com/webpack/webpack/pull/6314, the required options wer specified as boolean, glob, and array of globs.

**Does this PR introduce a breaking change?**

No.

**Other information**

### CODE REVIEW

**Comment by reergymerej** [comment]
I need some help here. This worked locally. I don’t know what’s happening here.

**Comment by reergymerej** [comment]
Thanks for the suggestion. I’ll give it a shot.

File: test/SideEffectsFlagPlugin.unittest.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+	it("should understand boolean values", () => {
+		SideEffectsFlagPlugin.moduleHasSideEffects("./foo/bar.js", true).should.eql(true);
+		SideEffectsFlagPlugin.moduleHasSideEffects("./foo/bar.js", false).should.eql(false);
+	});
+
+	it("should understand a glob", () => {
+		SideEffectsFlagPlugin.moduleHasSideEffects("./src/x/y/z.js", "./src/**/*.js").should.eql(true);
+		SideEffectsFlagPlugin.moduleHasSideEffects("./x.js", "./src/**/*.js").should.eql(false);
+		SideEffectsFlagPlugin.moduleHasSideEffects("./src/x/y/z.js", "**.js").should.eql(true);
+		// SideEffectsFlagPlugin.moduleHasSideEffects("./src/x/y/z.js", "*/src**").should.eql(true);
```
Comment (sokra) [inline test/SideEffectsFlagPlugin.unittest.js]: Either remove or include these lines. Don't leave code in comments.

**Comment by sokra** [comment]
continued and merged in https://github.com/webpack/webpack/pull/6536

**Pull request closed without merging**